pennylane/ops/cv.py

```diff
@@ -803,6 +807,63 @@ def _heisenberg_rep(p):
         return np.diag([-0.5, 0.5 / hbar, 0.5 / hbar])
 
 
+class TensorN(CVObservable):
+    r"""pennylane.ops.TensorN(wires)
+    The tensor product of the :class:`~.NumberOperator` acting on different wires.
+
+    If a single wire is defined, returns a :class:`~.NumberOperator` instance for convenient gradient computations.
+
+    When used with the :func:`~.expval` function, the expectation value
+    :math:`\langle \hat{n}_{i_0} \hat{n}_{i_1}\dots \hat{n}_{i_{N-1}}\rangle`
+    for a (sub)set of modes :math:`[i_0, i_1, \dots, i_{N-1}]` of the system is
+    returned.
+
+    **Example:**
+
+    >>> cv_obs = qml.TensorN(wires=[0, 1])
+    >>> cv_obs
+    TensorN(wires=[0, 1])
+    >>> cv_obs.ev_order is None
+    True
+
+    >>> cv_obs = qml.TensorN(wires=[1])
+    >>> cv_obs
+    NumberOperator(wires=[1])
+    >>> cv_obs.ev_order
+    2
+
+    **Details:**
+
+    * Number of wires: Any
+    * Number of parameters: 0
+
+    Args:
+        wires (Sequence[int] or int): the wire the operation acts on
+    """
+    num_wires = AnyWires
+    num_params = 0
+    par_domain = None
+    ev_order = None
+
+    def __new__(cls, *params, wires=None, do_queue=True):
+        # Custom definition for __new__ needed such that a NumberOperator can
+        # be returned when a single mode is defined
+
+        if wires is None:
+            wires = params[-1]
+            params = params[:-1]
+
+        if isinstance(wires, int) or len(wires) == 1:
+            instance = NumberOperator.__new__(NumberOperator)
+        else:
+            instance = super().__new__(cls)
+
+        # Need to call __init__ as it is not being invoked with a custom
+        # __new__ method
+        instance.__init__(*params, wires=wires, do_queue=do_queue)
+        return instance
```

Comment:
This looks great @antalszava!

A quick note, but as far as I'm aware, if `__new__` doesn't change the class, then `__init__` will be called automatically. So instantiation only needs to happen in `__new__` if the class changes, otherwise it will be called twice?

E.g.,
```python
class B:
    def __init__(self, *args):
        print("Init of B is called with args", *args)

class A:
    def __new__(cls, *args):
        print("Creating instance with args", *args)
        if len(args) == 2:
            return B(*args)
        return super(A, cls).__new__(cls)

    def __init__(self, *args):
        print("Init of A is called with args", *args)
```

Instantiating this class with 1 arg:
```python
>>> x = A(0)
Creating instance with args 0
Init of A is called with args 0
>>> print(x)
<__main__.A at 0x7f4c7c96cf60>
```
Instantiating this class with 2 args:
```python
>>> x = A(0, 1)
Creating instance with args 0 1
Init of B is called with args 0 1
>>> print(x)
<__main__.B at 0x7f4c7c96cf28>
```

Reply:
Great one, thanks!